File: leetcode 75/345. Reverse Vowels of a String/reverseVowels.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function(s) {
    let vowels = new Set("aeiouAEIOU");
    let l = 0;
    let r = s.length - 1;
    s = [...s];

    while (l < r) {
        if (vowels.has(s[l]) && vowels.has(s[r])) {
            [s[l], s[r]] = [s[r], s[l]];
            l++;
            r--;
        } else {
            if (!vowels.has(s[l])) l++;
            if (!vowels.has(s[r])) r--;
        }
    }

    return s.join('');
};

File: leetcode 75/345. Reverse Vowels of a String/reverseVowels.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        l = 0
        r = len(s) - 1
        s = list(s)

        while l < r:
            if s[l] in vowels and s[r] in vowels:
                s[l], s[r] = s[r], s[l]
                l += 1
                r -= 1
            else:
                if s[l] not in vowels:
                    l += 1
                if s[r] not in vowels:
                    r -= 1
        
        return ''.join(s)